json/targeting.json

```diff
@@ -0,0 +1,553 @@
+{
+  "$id": "https://flagd.dev/schema/targeting.json",
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "title": "flagd Targeting",
+  "description": "Defines targeting logic for flagd; a extension of JSONLogic, including purpose-built feature-flagging operations.",
+  "type": "object",
+  "$defs": {
+    "targeting": {
+      "title": "Targeting",
+      "description": "An expression returning a value which is coerced to a string to be used as a targeting key, or null (to fall back to defaultVariant). If targeting returns a value which is not a variant key, it's considered an error.",
+      "anyOf": [
+        {
+          "$comment": "we need this to support empty targeting",
+          "type": "object",
+          "additionalProperties": false,
+          "properties": {}
+        },
+        {
+          "$ref": "#/$defs/anyRule"
+        }
```

Comment:
Unfortunately, it looks like this messes with VSCodes intellisense. Valid operators don't autocomplete until the empty targeting support section is commented out.

Reply:
I think this is more about how vscode in particular does auto-completion and not about JSON schema. It seems that since `{}` is valid there's not much I can do (though I did try a few things to make vscode behave as we'd want). Frankly I'm not sure even if we got it working that it would stay working...

I think we can either remove the `{}` support (and optionally remove it from flagd also?) or accept that vscode auto-complete won't work for the top-level rule (it works for the rest since `{}` is only acceptable at the root).

If we just remove the `{}` from the schema, it might be acceptable... whether or not we actually remove the support in flagd. I'm not really sure there's much value in supporting it officially. What are your thoghts?

Reply:
@beeme1mr thinking about it more, I believe we should just not consider empty targeting valid from the schema perspective. We can keep the actual flags implementations as they are. Empty targeting just returns the defaultValue anyway.

Reply:
I was also able to get it working if I change "anyRule" to use "anyOf" instead of "oneOf" but I'm not sure what impact that may have.

Reply:
I'll set if that has any negative impacts.

Reply:
Actually that seems to work without issue... empty targeting `{}` is still valid, and vscode seems to auto-complete, and all the validation continues to work otherwise. I don't really understand why vscode can't auto-complete with "oneOf" as any rule, but I think there's no downside to this so :shrug: